### DISCUSSION TITLE
Importing Blender model to Stride guide

### DISCUSSION THREAD

**Discussion by Basewq** [stride3d/stride#2414] [General]
This discussion is help people importing models from Blender to Stride.
If you have found an alternative or easier way, feel free to contribute your steps.

As far as I'm aware importing fbx files is the easiest option.
The following guide was tested with Stride `4.2.0.2149` and Blender `4.0.2`
### Importing static mesh with texture(s) with fbx export

0. Create/save your blender file somewhere, and create your mesh.

1. Ensure the mesh has a material set to the mesh.

- Select the mesh in the Scene Collection -> Materials tab
![ss1](https://github.com/user-attachments/assets/8e8dda9f-37bb-4861-a99b-b7e6b209661e)

2. Make the material a reference the texture

- Select the 'Shading' tab at the top.
- Select the mesh in the Scene Collection
- From the Base Color yellow circle in the 'Principled BSDF' node, mouse drag out to the left.
- Search/select 'Image texture'
![ss2](https://github.com/user-attachments/assets/b8e52bb2-2f48-4bd9-800a-4ec3e31985a3)

- Find/create a texture to use, and copy it over the same folder as your `.blend` file.
- In the 'Image Texture' node, click Open and select your texture from the previous step.
![ss3a](https://github.com/user-attachments/assets/fa494a26-f674-466e-85cf-d9f2b6005789)

3. (Optional step) View/edit your texture UV

- Depending on your version of Blender, sometimes Blender is buggy viewing the material, the following settings may help make the texture visible.
- Edit your mesh UV with the 'UV Editing' tab at the top (beyond the scope of this guide)
![ss3b](https://github.com/user-attachments/assets/7994377d-0f07-469e-b999-6f989337bc8b)

4. Export the model

- Select the mesh in the Scene Collection
- File -> Export -> FBX
![ss4](https://github.com/user-attachments/assets/b7b1d304-e971-4217-bc7f-054448fee2b4)

- Select 'Path Mode' `Copy`
- The icon next to the Path Mode options is a 'Embed Textures' option. Either options are valid, but you will need to note how they differ: If *enabled*, the textures are embedded into the fbx (larger file size), but effectively a single file. If *not* enabled the textures are in a `*.fbm` sub-folder - this will require an extra step in the next step.
- As a habit, it is a good idea to tick 'Limit to Selected Objects'
- Selected Object Types should be restricted to Mesh (and Armature if you intend to add animation)
![ss4b](https://github.com/user-attachments/assets/b5c2680f-38ef-4b7f-8408-8529b5fdd404)

5. Import the fbx into Stride

- Move the fbx into the Game project's Resources folder (or sub-folder).
- If you have *not* selected 'Embed Textures', open the `*.fbm` folder and move the files to the **same** folder as the fbx file.
- Drag and drop the fbx into the Asset view window (or Add asset -> Model -> 3D Model)
- Ensure Import materials & Import textures are ticked
![ss5](https://github.com/user-attachments/assets/8c5fd7da-c3df-4ac0-b975-9d536ded7a92)

- **IMPORTANT SIDE NOTE:** It is extremely important to use the simplest material setup, as Blender does not always detect texture references for export, eg. the following material node settings can be detect referenced textures
![ss5_mat](https://github.com/user-attachments/assets/2a10deb8-bace-4082-9d82-5e7d3259052f)
Note how for the normal map, the normal map image must be connected to a 'Normal Map' -> Color node, otherwise Blender will not export the normal map image.

**Comment by Basewq** [comment]
The following guide was tested with Stride `4.2.0.2149` and Blender `4.0.2`
### Importing animated model with fbx export
This guide can be considered a continuation of the 'Importing static mesh with texture(s) with fbx export' guide.


0. Assuming you've already created a mesh, create an Armature.

![arm_ss1](https://github.com/user-attachments/assets/2aaf805c-00d2-4f64-8c1a-b661ac2c44a1)

Optional - You can move the Armature around in the Scene Collection, which should not affect how you export it.
![arm_ss2](https://github.com/user-attachments/assets/b53ecdcf-94b4-4f50-a603-7885df9b22da)

0b. Select the Armature in the Scene Collection, change to Edit Mode (or press `Tab` key) and add/modify the bones as you want.
Refer to the Blender docs for more info: https://docs.blender.org/manual/en/4.0/animation/armatures/bones/editing/introduction.html
(Hide the mesh with the eye icon if it's in the way)
![arm_ss4](https://github.com/user-attachments/assets/bb20e3ba-c633-4f2a-a019-182c617cdf50)


0c. Parent the mesh to the Armature. Refer to the Blender docs for more info: https://docs.blender.org/manual/en/4.0/animation/armatures/skinning/parenting.html
Quickest option is the following:
In the Scene Collection sub-window: Left click mesh, then Ctrl+Left click Armature (Make sure you select them in this specific order)
Ctrl+P -> Armature Deform (With Automatic Weights)

![arm_ss5](https://github.com/user-attachments/assets/183cc221-27d8-4b28-97e2-e7b3c36ec419)

Notice how now the Mesh object will move be a child of the Armature object:

![arm_ss6](https://github.com/user-attachments/assets/2d91051f-d1a8-41ca-b5be-00fbd78e8d02)


2. Create your animation

- Go to the Animation tab
- Select your Armature in the Scene Collection
- Ensure 'Armature Overlays' is enabled if you want to see the bone over your mesh

![arm_ss7](https://github.com/user-attachments/assets/49766fb4-9bbb-4cd3-be74-c9db17c0156a)


- On the bottom half of the Animation tab, change Editor context from 'Dope Sheet' to 'Action Editor'
- Press New for a new animation.
![arm_ss8](https://github.com/user-attachments/assets/b7f29932-4417-4818-b2ac-d5e062f97dcf)

- Rename the animation to whatever name you want
![arm_ss9](https://github.com/user-attachments/assets/64205a94-2131-481c-a6f8-01aa425451b9)

- **Important Note:** Do not push the action to the NLA stack. All animations should just be `Action`s.


3. Edit your animation

- Refer to the Blender docs for a more in depth guide https://docs.blender.org/manual/en/4.0/editors/dope_sheet/action.html

Quick summary is the following:
- Move the blue marker along the timeline where you want a new pose
- Change the bone(s) to your desired pose.
- Select the changed bones (or select all bones to be sure), press `I` key and insert your Keyframe (eg. Location & Rotation)
- Don't forget to set your start/end frames to ensure the animation will loop or stop properly.
![arm_ss10](https://github.com/user-attachments/assets/06375eef-eacc-4eb9-bff3-46bce234495c)


- Optional: Creating an additional Action (animation) is highlighted below:
![arm_ss10b](https://github.com/user-attachments/assets/52f2dedf-cf91-43c5-943d-acbe01cd2e1d)

(The icon left of the Action name is how you can select an existing Action that you've already made)


4. Export the model/animation

- Before exporting, reset all pose transform back to its resting pose, otherwise the fbx exporter passes incorrect default transforms:
- Select the armature, change to Pose Mode, select all bones (key 'A') -> Pose -> Clear Transform -> All
![arm_ss11b](https://github.com/user-attachments/assets/51f668e7-3146-444b-994e-524cd8b417ec)


- Select the mesh AND armature in the Scene Collection - You can do this by left click Mesh -> Ctrl+left click Armature, alternatively you can right click the Armature and choose Select Hierarchy
![arm_ss11](https://github.com/user-attachments/assets/0df6f57e-9a1c-46e9-810c-6355632f25cb)

- Like the previous guide, File -> Export -> FBX
- The key settings are similar to the previous guide:
- Use 'Path Mode' `Copy` to ensure the textures copied
- Ensure Object Types: **Mesh** and **Armature** are selected (use shift left click to select multiple Object Types)
- Ensure 'Bake Animations' -> 'All Actions' is selected.
- The other settings are dependent on your use case.

![arm_ss12](https://github.com/user-attachments/assets/529bdf5f-3e96-41a8-9962-11e7b6c6fd1a)


5. Import the fbx into Stride

- Move the fbx into the Game project's Resources folder (or sub-folder).
- If you have *not* selected 'Embed Textures', open the `*.fbm` folder and move the files to the **same** folder as the fbx file.
- Drag and drop the fbx into the Asset view window (or Add asset -> Model -> 3D Model)
- Ensure Import Animations & Import skeleton are ticked
![arm_ss13](https://github.com/user-attachments/assets/850299f4-0738-4446-ba9e-56ebd2040ada)

**Reply by VaclavElias** [reply]
Thanks for this. Let's hope we can get some feedback 🙂

**Reply by mihai-ene-public** [reply]
This kind of tutorials are very helpful and they should be in the manual.

Also the kind of tutorials that you'd use in a material with PBR textures would be also helpful, for example how would you create a stride material with a texture from here: https://polyhaven.com/a/laminate_floor_02. In my own example I'm having a hard time in using the roughness texture since it is not very clear how should this be used in a stride material.

Please keep going :)

**Reply by VaclavElias** [reply]
That's correct @mihai-ene-public. We just need a community which will write these. Docs is also community driven. Once this is reviewed, I will move it to our docs 🙂

**Reply by morenostef** [reply]
Hi,when i import fbx smooth models,following the rules, i find that the mesh is not smooth but it's flat.
This problem doesn't arise for obj object,if they are smooth in Blender (version 4.2) ,they stay smooth they once imported,
Any suggestions?

**Reply by Basewq** [reply]
You will probably need to clarify your problem more since there's not enough info to get from this.
I can only guess the normals are not exported correctly.

**Reply by morenostef** [reply]
Hi. Normals seems ok. I think it could be a problem of Blender,because loading models created with previous Blender versions, everything works @Basewq

**Comment by Doprez** [comment]
@VaclavElias tagging that issue here since this could definitely turn into an official guide at some point.

https://github.com/stride3d/stride/issues/2433#issuecomment-2339238624
